### PR TITLE
fix(challenges): Regex Asterisk Challenge

### LINKED ISSUE #17872: Regular Expressions - asterisk
#### Describe your problem and - if possible - how to reproduce it
I find this challenge's instructions pretty confusing. I made the hard part bold.

> The last challenge used the plus + sign to look for characters that occur one or more times. There's also an option that matches characters that occur zero or more times.
> 
> The character to do this is the asterisk or star: *.
> 
>     let soccerWord = "gooooooooal!";
>     let gPhrase = "gut feeling";
>     let oPhrase = "over the moon";
>     let goRegex = /go*/;
>     soccerWord.match(goRegex); // Returns ["goooooooo"]
>     gPhrase.match(goRegex); // Returns ["g"]
>     oPhrase.match(goRegex); // Returns null
> 
> **_Create a regex chewieRegex that uses the * character to match all the upper and lower"a" characters in chewieQuote._** Your regex does not need flags, and it should not match any of the other quotes.

somehow this worked:
```
let chewieQuote = "Aaaaaaaaaaaaaaaarrrgh!";
let chewieRegex = /Aa*/; // Change this line
let result = chewieQuote.match(chewieRegex);
```
Here are tests:

1. Your regex chewieRegex should use the * character to match zero or more a characters.
2. Your regex chewieRegex should match 16 characters.
3. Your regex should match "Aaaaaaaaaaaaaaaa".
4. Your regex should not match any characters in "He made a fair move. Screaming about it can't help you."
5. Your regex should not match any characters in "Let him have it. It's not wise to upset a Wookiee."

4th test is especially confusing. there are lowercase 'a', why should I _not_(added after edit) match them?
#### Add a Link to the page with the problem
https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/regular-expressions/match-characters-that-occur-zero-or-more-times

### PR DESCRIPTION
**Pull request by scissorsneedfoodtoo**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Fixes description and adds several tests to close a long standing issue with this challenge's description.

Closes #17872

### CODE REVIEW

**Comment by thecodingaviator** [comment]
@scissorsneedfoodtoo I'd appreciate if you could also remove the `testString` from the `testString` and only let the assert be there. Due to new changes, only the `text` string would be used everywhere now with reserved exceptions

**Comment by scissorsneedfoodtoo** [comment]
@thecodingaviator, thanks, I didn't realize that. So with the recent changes the tests should be:

```
  - text: Your regex should not match <code>"aaaa"</code>.
    testString: assert(!chewieRegex.test('aaaa'));
```

Is that right?

**Comment by thecodingaviator** [comment]
@scissorsneedfoodtoo IMO that is right, but since I have never tried it, @RandellDawson would be a better person to consult that with

**Comment by RandellDawson** [comment]
@scissorsneedfoodtoo , @thecodingaviator This is not something we should require users to do on a PR, but it is now an option we can implement.  If we expect this, I believe it should be in the `How to Work on Guide Challenges` documentation and then only apply to PRs after that policy is implemented.  

In general, we should not hold up PRs for extra stuff which goes beyond the original scope of the PR or issue the PR attempts to close.  If other fixes are needed beyond that scope, I believe they should be put on a new issue, so a discussion can take place and a separate PR created to address concerns.  

I see many good PRs held up from a merge, because we find other issues beyond the original scope.  Many times these PRs go stale, which seems unnecessary IMHO.

**FYI** - Removing the `message` argument from the assert only applies when there is a single `assert` in a `testString` per `text`.  There are some challenges which use `throw` and may only have one `text`, but multiple asserts.

**Comment by scissorsneedfoodtoo** [comment]
@RandellDawson, thanks for clearing that up. Makes sense and I'll keep it in mind going forward.

@thecodingaviator, I don't mind removing the message arguments here, even if if it's a bit outside the scope of this PR. I tested it yesterday and it seemed to work fine.

**Comment by scissorsneedfoodtoo** [comment]
@thecodingaviator, Sorry for the delay! I removed the unnecessary string arguments from testString. Could you take another look when you can?

**Comment by thecodingaviator** [comment]
@scissorsneedfoodtoo See this, it passes:
image

But just by changing the top line, it fails the same RegEx, we should be checking for the regex but without using the `chewieQuote`:
image

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/match-characters-that-occur-zero-or-more-times.english.md
Comment (thecodingaviator) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/match-characters-that-occur-zero-or-more-times.english.md]:
```suggestion
    testString: assert(chewieRegex.test('Aaaaaaaaaaaaaaaarrrgh!'));
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/match-characters-that-occur-zero-or-more-times.english.md
Comment (thecodingaviator) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/match-characters-that-occur-zero-or-more-times.english.md]:
```suggestion
    testString: assert(chewieRegex.test('Aaaaaaaaaaaaaaaa'));
```

**Review by thecodingaviator — DISMISSED** [review]
This should fix it IMO. Has been tested locally
image

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/match-characters-that-occur-zero-or-more-times.english.md
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/match-characters-that-occur-zero-or-more-times.english.md]: I might be wrong about this, but I believe this change would cause incorrect answers to pass these tests. `test()` just checks whether a pattern is found anywhere in the string, so the following would work:

`let chewieRegex = /A/;`
image

let chewieRegex = /a/;
image

Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/match-characters-that-occur-zero-or-more-times.english.md]: What we could do is change the tests to `'Aaaaaaaaaaaaaaaa'.match(chewieQuote)`, or some variation of that to make sure we're just testing the regex itself. I'm not sure if that's the best move, though, since it might give users the wrong impression about how regex works if we're hard coding some values behind the scenes.

If the user changes the first string, then plays with different regex patterns until they pass the challenge, they could be confused by the textString that says their regex should match 16 characters.

Sorry, I'm not sure if I explained that well. Still, I'm open to any suggestions!

Comment (thecodingaviator) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/match-characters-that-occur-zero-or-more-times.english.md]: You're right. Hmm. Interesting problem.

@RandellDawson Do you have any ides how we should go by tackling this?

Comment (RandellDawson) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/match-characters-that-occur-zero-or-more-times.english.md]: I need to review it again.  Give me some time.

**Comment by RandellDawson** [comment]
@scissorsneedfoodtoo Working with @thecodingaviator, we came up with the following verbiage for both the Instructions and the actual tests.  We believe this should resolve the original issue and make the instructions clearer.

**Instructions**
>For this challenge, <code>chewieQuote</code> has been initialzied as "Aaaaaaaaaaaaaaaarrrgh!" behind the scenes. Create a regex <code>chewieRegex</code> that uses the <code>*</code> character to match an uppercase <code>"A"</code> character immediately followed by zero or more lowercase <code>"a"</code> characters in <code>chewieQuote</code>. Your regex does not need flags or character classes, and it should not match any of the quotes.

**Tests**
```yml
tests:
  - text: Your regex <code>chewieRegex</code> should use the <code>*</code> character to match zero or more <code>a</code> characters.
    testString: assert(/\*/.test(chewieRegex.source));
  - text: Your regex should match <code>"A"</code> in <code>chewieQuote</code>.
    testString: assert(result[0][0] === 'A');
  - text: Your regex should match only <code>"aaaa"</code> when immediately preceeded by <code>"A"</code> in <code>chewieQuote</code>.
    testString: assert(result[0].slice(1, 5) === 'aaaa');
  - text: Your regex should match <code>"Aaaaaaaaaaaaaaaa"</code> in <code>chewieQuote</code>.
    testString: assert(result[0] === 'Aaaaaaaaaaaaaaaa');
  - text: Your regex <code>chewieRegex</code> should match 16 characters in <code>chewieQuote</code>.
    testString: assert(result[0].length === 16);
  - text: Your regex should not match any characters in "He made a fair move. Screaming about it can't help you."
    testString: assert(!"He made a fair move. Screaming about it can't help you.".match(chewieRegex));
  - text: Your regex should not match any characters in "Let him have it. It's not wise to upset a Wookiee."
    testString: assert(!"Let him have it. It's not wise to upset a Wookiee.".match(chewieRegex));

```

Please let us know if you think these changes would work.

**Comment by scissorsneedfoodtoo** [comment]
@RandellDawson and @thecodingaviator, just tested these changes and everything LGTM. Thank you both for all your help with this! The new wording for the instructions and tests read very well, and moving the `chewieQuote` variable to the before test section to prevent accidental changes was a great idea.

Please go ahead an commit them when you get a chance

**Comment by RandellDawson** [comment]
@scissorsneedfoodtoo After looking at my suggestion a bit more, we decided it no longer made sense to have the following test, so I removed it.

```yml
  - text: Your regex should match only <code>"aaaa"</code> when immediately preceeded by <code>"A"</code> in <code>chewieQuote</code>.
    testString: assert(result[0].slice(1, 5) === 'aaaa');
```

**Comment by scissorsneedfoodtoo** [comment]
@RandellDawson, that sounds good. Seems redundant with all the other tests.

**Review by RandellDawson — APPROVED** [review]
Let's get this merged.  Need another reviewer to approve and merge.

**Review by thecodingaviator — APPROVED** [review]
LGTM! 👍

**Comment by thecodingaviator** [comment]
Thank you for your contribution to the page! 👍
We're happy to accept these changes, and look forward to future contributions. 📝

**Comment by scissorsneedfoodtoo** [comment]
Thank you for the merge, and for all of your help with this @thecodingaviator and @RandellDawson!

**Pull request merged**
